Makefile

```diff
@@ -1,16 +1,147 @@
-.PHONY: docker-build
+# Copyright 2023 K8sgpt AI. All rights reserved.
+# Use of this source code is governed by a MIT style
+# license that can be found in the LICENSE file.
+
+# ==============================================================================
+# define the default goal
+#
+ROOT_PACKAGE=github.com/k8sgpt-ai/k8sgpt
+
+SHELL := /bin/bash
+DIRS=$(shell ls)
+GO=go
+
+.DEFAULT_GOAL := help
+
+# include the common makefile
+COMMON_SELF_DIR := $(dir $(lastword $(MAKEFILE_LIST)))
+# ROOT_DIR: root directory of the code base
+ifeq ($(origin ROOT_DIR),undefined)
+ROOT_DIR := $(abspath $(shell cd $(COMMON_SELF_DIR)/. && pwd -P))
+endif
+# OUTPUT_DIR: The directory where the build output is stored.
+ifeq ($(origin OUTPUT_DIR),undefined)
+OUTPUT_DIR := $(ROOT_DIR)/bin
+$(shell mkdir -p $(OUTPUT_DIR))
+endif
+
+ifeq ($(origin VERSION), undefined)
+VERSION := $(shell git describe --abbrev=0 --dirty --always --tags | sed 's/-/./g')
+endif
+
+# Check if the tree is dirty. default to dirty(maybe u should commit?)
+GIT_TREE_STATE:="dirty"
+ifeq (, $(shell git status --porcelain 2>/dev/null))
+	GIT_TREE_STATE="clean"
+endif
+GIT_COMMIT:=$(shell git rev-parse HEAD)
+
 IMG ?= ghcr.io/k8sgpt-ai/k8sgpt:latest
 
+BUILDFILE = "./main.go"
+BUILDAPP = "$(OUTPUT_DIR)/k8sgpt"
+
+.PHONY: all
+all: tidy add-copyright lint cover build
+
+# ==============================================================================
+# Targets
+
+## build: Build binaries by default
+.PHONY: build
+build: 
+	@echo "$(shell go version)"
+	@echo "===========> Building binary $(BUILDAPP) *[Git Info]: $(VERSION)-$(GIT_COMMIT)"
+	@export CGO_ENABLED=0 && go build -o $(BUILDAPP) -ldflags '-s -w' $(BUILDFILE)
+
+## tidy: tidy go.mod
+.PHONY: tidy
+tidy:
+	@$(GO) mod tidy
+
+## deploy: Deploy k8sgpt
+.PHONY: deploy
 deploy:
+	@echo "===========> Deploying k8sgpt"
+	@$(call funcsecret)
+
+## undeploy: Undeploy k8sgpt
+.PHONY: undeploy
+undeploy:
+	@echo "===========> Undeploying k8sgpt"
+	kubectl delete secret ai-backend-secret --namespace=k8sgpt
+	kubectl delete -f container/manifests
+	kubectl delete ns k8sgpt
+
+## docker-build: Build docker image
+.PHONY: docker-build
+docker-build:
+	@echo "===========> Building docker image"
+	docker buildx build --build-arg=VERSION="$$(git describe --tags --abbrev=0)" --build-arg=COMMIT="$$(git rev-parse --short HEAD)" --build-arg DATE="$$(date +%FT%TZ)" --platform="linux/amd64,linux/arm64" -t ${IMG} -f container/Dockerfile . --push
+
+
+## fmt: Run go fmt against code.
+.PHONY: fmt
+fmt:
+	@$(GO) fmt ./...
+
+## vet: Run go vet against code.
+.PHONY: vet
+vet:
+	@$(GO) vet ./...
+
+## lint: Run go lint against code.
+.PHONY: lint
+lint:
+	@golangci-lint run -v ./...
```

Comment:
Shouldn't verify `golangci-lint` is installed ?

```bash
## lint: Run go lint against code.
.PHONY: lint
lint:
	@if [ -x "$$(command -v golangci-lint)" ]; then \
		golangci-lint run -v ./... ; \
	else \
		echo "golangci-lint is not installed. Please install it using 'cd ../ && go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.52.2'" ; \
	fi
```

Reply:
Yes, I think the installation should be verified and even judged, but I think one makefile is now too much. I would like to extend the script directory and divide it into makefile subdirectories